File: src/api/security-check-custom-password.ts
import { RequestFN } from "~/core/request-function";
import { DoubleAuthServerAction, SessionHandle } from "~/models";
import { aesKeys } from "./private/keys";
import { AES } from "~/api/private/aes";

export const securityCheckCustomPassword = async (session: SessionHandle, newPassword: string): Promise<boolean> => {
  const keys = aesKeys(session);

  const request = new RequestFN(session, "SecurisationCompteDoubleAuth", {
    donnees: {
      action: DoubleAuthServerAction.csch_VerifierMotDePassePersonnalise,
      nouveauMDP: AES.encrypt(newPassword, keys.key, keys.iv)
    }
  });

  const response = await request.send();
  return response.data.donnees.result;
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/api/security-check-custom-password.kt
package ink.literate.pawnote.api

import ink.literate.pawnote.api.private.AES
import ink.literate.pawnote.api.private.aesKeys
import ink.literate.pawnote.core.RequestFN
import ink.literate.pawnote.models.DoubleAuthServerAction
import ink.literate.pawnote.models.SessionHandle
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*

suspend fun securityCheckCustomPassword(session: SessionHandle, newPassword: String): Boolean {
  val keys = aesKeys(session.information)

  val request =
      RequestFN(
          session.information,
          "SecurisationCompteDoubleAuth",
          Json.encodeToString(
              buildJsonObject {
                putJsonObject("donnees") {
                  put("action", DoubleAuthServerAction.csch_VerifierMotDePassePersonnalise.code)
                  put("nouveauMDP", AES.encrypt(newPassword.toByteArray(), keys.key, keys.iv))
                }
              }))

  val response = request.send()
  return Json.parseToJsonElement(response.data)
      .jsonObject["donnees"]!!
      .jsonObject["result"]!!
      .jsonPrimitive
      .boolean
}
